cookbooks/bcpc/recipes/serial_console.rb

```diff
@@ -0,0 +1,71 @@
+# declare a placeholder template
+template "/etc/init/ttySx.conf" do
+    source "ttySx.erb"
+    owner "root"
+    group "root"
+    mode 00644
+    action :nothing
+end
+
+# declare a placeholder service
+service "ttySx_svc" do
+   action :nothing
+   provider Chef::Provider::Service::Upstart
+end
+
+ruby_block "discover_serial" do
+    block do
+        def numeric?(data)
+                Float(data) != nil rescue false
+        end
+        # we are not interested in USB or any other serial at the moment
+        # Sample file format where the first column
+        # number x corresponds to ttySx
+        # serinfo:1.0 driver revision:
+        # 0: uart:16550A port:000003F8 irq:4 tx:165 rx:0 CTS|DSR|CD
+        # 1: uart:unknown port:000002F8 irq:3
+        # 2: uart:unknown port:000003E8 irq:4
+        serial_dev_type = "ttyS"
+        serial_driver_info = "/proc/tty/driver/serial"
+        File.open(serial_driver_info, "r") do |serials|
+            serials.each_line do |line|
+                if line.start_with? "serinfo:"
+                    next
+                end
+                line_parts = line.split()
+                uart = line_parts[1]
+                if not uart.include? "uart:unknown"
+                    port_assignment = line_parts[0]
+                    port_number = port_assignment.chop
+                    # get device baud we expect
+                    # root@bcpc-vm2:/proc/tty/driver# stty -F /dev/ttyS0 speed
+                    # 115200
+                    device_baud = `stty -F /dev/#{serial_dev_type}#{port_number} speed`
```

Comment:
The OS command can be executed using Mixlib::ShellOut in more controlled way and better error handling. One of the example is [`namenode_master.rb`](https://github.com/amithkanand/chef-bach/blob/master/cookbooks/bcpc-hadoop/recipes/namenode_master.rb#L126-L136)